x/merkledb/benchmarks/benchmark.go

```diff
@@ -0,0 +1,404 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package main
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/binary"
+	"fmt"
+	"net/http"
+	"os"
+	"path"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/spf13/pflag"
+
+	"github.com/ava-labs/avalanchego/database/leveldb"
+	"github.com/ava-labs/avalanchego/trace"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/utils/units"
+	"github.com/ava-labs/avalanchego/x/merkledb"
+)
+
+const (
+	defaultDatabaseEntries    = 1000000000 // 1B
+	databaseCreationBatchSize = 10000      // 10k
+	databaseRunningBatchSize  = 2500       // 2.5k
+	databaseRunningUpdateSize = 5000       // 5k
+	defaultMetricsPort        = 3000
+)
+
+var (
+	databaseEntries = pflag.Uint64("n", defaultDatabaseEntries, "number of database entries")
+	httpMetricPort  = pflag.Uint64("p", defaultMetricsPort, "default metrics port")
+	verbose         = pflag.Bool("v", false, "verbose")
+
+	insertsCounter = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "merkledb_bench",
+		Name:      "insert_counter",
+		Help:      "Total number of inserts",
+	})
+	deletesCounter = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "merkledb_bench",
+		Name:      "deletes_counter",
+		Help:      "Total number of deletes",
+	})
+	updatesCounter = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "merkledb_bench",
+		Name:      "updates_counter",
+		Help:      "Total number of updates",
+	})
+	batchCounter = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: "merkledb_bench",
+		Name:      "batch",
+		Help:      "Total number of batches written",
+	})
+	deleteRate = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "merkledb_bench",
+		Name:      "entry_delete_rate",
+		Help:      "The rate at which elements are deleted",
+	})
+	updateRate = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "merkledb_bench",
+		Name:      "entry_update_rate",
+		Help:      "The rate at which elements are updated",
+	})
+	insertRate = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "merkledb_bench",
+		Name:      "entry_insert_rate",
+		Help:      "The rate at which elements are inserted",
+	})
+	batchWriteRate = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "merkledb_bench",
+		Name:      "batch_write_rate",
+		Help:      "The rate at which the batch was written",
+	})
```

Comment:
We should not be calculating the rates in the benchmark. The prometheus server should do this based on the counts.
```suggestion
```

Reply:
I believe that it won't generate accurate results, since we're mixing batch writing and put in the same sequence.
I've included both the counter and the rate metrics so that we can get both numbers in the grafana.